### ISSUE TITLE
TFlite Poor Performance on Pixel 3 CPU when channel not multiple of 4

### ISSUE THREAD

**Issue by Makkiy** [tensorflow/tensorflow#46002]
**System information**
- Have I written custom code: Only to generate conv2d tflite models
- Host OS Platform and Distribution: Linux Ubuntu 16.04
- Mobile device: Pixel 3
- TensorFlow installed from (source or binary): [native benchmark binary](https://www.tensorflow.org/lite/performance/measurement#native_benchmark_binary) nightly pre-built (no Flex delegate) android_aarch64
- Host TensorFlow version: 2.3.0
- Mobile device TensorFlow version: nightly
- Python version: 3.8.3
- Bazel version (if compiling from source): n/a
- GCC/Compiler version (if compiling from source): n/a
- CUDA/cuDNN version: n/a
- GPU model and memory: n/a

**Describe the current behavior**
The tflite conv2d op performs poorly on Pixel 3 CPU when the input or output channel numbers are not multiples of 4. Selective inference latency measured on Pixel 3 is shown in the following table to support 3 major observations:

|In/Out Channel|Latency|In/Out Channel|Latency|In/Out Channel|Latency|
|-|-|-|-|-|-|
|16/14|21760.5|30/43|137299.0|4/44|14541.9|
|16/15|24486.5|31/43|166752.0|44/4|38141.8|
|16/16|21479.2|32/43|86515.3|11/43|67771.0|
|16/17|24136.2|33/43|156961.0|43/11|72423.9|
|16/18|26734.9|34/43|152460.0|22/30|66271.4|
|44/37|26193.0|35/43|158265.0|30/22|77171.4|
|44/64|124600.0|36/43|112036.0|

1. The first two columns show that with a fixed input channel, output channels of no multiples of 4 (no4x) perform even worse than those of more channels but of multiples of 4 (4x). Notably, 44/37 has a larger latency than 44/64.
2. Similarly, the second two columns show that with a fixed output channel, no4x input channels perform worse than 4x input channels. Notably, 30/43 has a larger latency than 36/43.
3. The last two columns show that input and output channels are not created equal. Regardless of 4x or no4x, input channels turn to induce greater latency than output channels.

**Describe the expected behavior**
[TFlite guide on GPU](https://www.tensorflow.org/lite/performance/gpu_advanced#tips_and_tricks) mentions:

> On a GPU, tensor data is sliced into 4-channels. Thus, a computation on a tensor of shape [B, H, W, 5] will perform about the same on a tensor of shape [B, H, W, 8], but significantly worse than [B, H, W, 4].

Our questions:

1. Should CPU observe the same behavior as the above-described GPU case?
2. The above example seems to indicate that the latency of a no4x channel should perform about the same as when it is rounded up to the next 4x, but in our measurements, it is not the case. no4x can sometimes perform worse than much larger 4x. Is there any reasoning behind this?
3. In our third observation, input channel impacts latency more than output channel. Is there any reasoning behind this?

**Standalone code to reproduce the issue**
The conv2d models used for the benchmark are defined with keras and converted to tflite with the following code:
```
# TF model generation
model = Sequential()
model.add(Conv2D(output_channel, (3, 3), padding="same"))
model.build(input_shape=(1, 180, 180, input_channel))
shutil.rmtree("./temp")
save_model(model, "./temp")
converter = tf.lite.TFLiteConverter.from_saved_model(saved_model_dir="./temp", signature_keys=['serving_default'])
converter.optimizations = [tf.lite.Optimize.DEFAULT]
converter.experimental_new_converter = True
tflite_model = converter.convert()
tf_filename = 'b{}_ic{}_oc{}_ih{}_iw{}_tf.tflite'.format(1, input_channel, output_channel, 180, 180)
with open(model_dir + tf_filename, 'wb') as f:
    f.write(tflite_model)
```
Where `input_channel` and `output_channel` are variables, batch size is fixed to 1, input shape is fixed to 180x180, kernel size is 3x3, and stride is 1. In this way, the generated tflite models contain only a single conv2d operator with various input and output channel numbers. The models are then loaded to Pixel 3 phones for the benchmark.

**Other info / logs**
The inference latency was measured using the [native benchmark binary](https://www.tensorflow.org/lite/performance/measurement#native_benchmark_binary) nightly pre-built (no Flex delegate) android_aarch64 with the following command:
`adb shell taskset f0 /data/local/tmp/android_aarch64_benchmark_model --graph=/data/local/tmp/{model}.tflite --num_threads=1 --num_runs=100`
In the report generated by the benchmark binary, `inference (avg)` is kept as the inference latency of a conv2d model. For example, when the report contains
`Inference timings in us: Init: 8241, First inference: 4084, Warmup (avg): 1584.61, Inference (avg): 1573.36`
1573.36 is kept as the latency for the model.

**Comment by licj15** [comment]
I found similar things on my side, not sure if there is any comment from TF?

**Comment by pkgoogle** [comment]
Hi @Makkiy, can you try with the most recent version and let us know if this is still an issue for you?
